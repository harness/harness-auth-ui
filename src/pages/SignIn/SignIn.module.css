/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.signin {
  & .header {
    display: flex;
    margin-bottom: var(--spacing-huge);
    align-items: center;
    font-size: var(--font-size-medium);
  }

  & .forgotLink {
    position: absolute;
    right: 0;
  }

  & .title {
    font-size: 32px;
    font-weight: 600;
    margin-bottom: var(--spacing-small);
  }

  & .subtitle {
    font-size: var(--font-size-medium);
    margin-bottom: var(--spacing-huge);
    color: var(--dark-400);
  }

  & .footer {
    margin-top: var(--spacing-huge);
  }

  & .center {
    text-align: center;
  }
}
