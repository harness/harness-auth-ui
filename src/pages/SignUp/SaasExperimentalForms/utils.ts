/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import CDLogo from "static/images/cd-logo.svg";
import CILogo from "static/images/ci-logo.svg";
import CCMLogo from "static/images/ccm-logo.svg";
import FFLogo from "static/images/ff-logo.svg";
import ChaosLogo from "static/images/chaos-logo.svg";
import HarnessLogo from "static/images/logo.svg";

import PathCD from "static/images/path-cd.svg";
import PathCI from "static/images/path-ci.svg";
import PathCCM from "static/images/path-ccm.svg";
import PathFF from "static/images/path-ff.svg";
import PathChaos from "static/images/path-chaos.svg";
import PathHarness from "static/images/path-harness.svg";

interface ModuleDetailsMap {
  [key: string]: ModuleDetail;
}

interface ModuleDetail {
  logo: string;
  pathImg: string;
  title: string;
  tagLine: string;
  valueProp: string;
  callout: string;
  intro: string;
  cta: string;
  color: string;
}
const MODULE_TO_DETAILS_MAP: ModuleDetailsMap = {
  cd: {
    logo: CDLogo,
    pathImg: PathCD,
    title: "Continuos Delivery",
    tagLine: "Blazing Fast Deployment Pipelines in Minutes",
    valueProp:
      "Self-Service Continuous Delivery that enables engineers to deploy on-demand, without a single script",
    callout:
      "Truatsed by 1000s of Developers, driving 17 Million deployments in last 12 months",
    intro:
      "Define Pipelines that enable you to improve your deployment times, manage all in one place",
    cta: "Let's Deploy",
    color: "#5FB34E"
  },
  chaos: {
    logo: ChaosLogo,
    pathImg: PathChaos,
    title: "Chaos Engineering",
    tagLine: "Blazing Fast Deployment Pipelines in Minutes",
    valueProp:
      "Self-Service Continuous Delivery that enables engineers to deploy on-demand, without a single script",
    callout:
      "Trusted by 1000s of Developers, driving 17 Million deployments in last 12 months",
    intro:
      "Define Pipelines that enable you to improve your deployment times, manage all in one place",
    cta: "Let's Deploy",
    color: "#FF006A"
  },
  ci: {
    logo: CILogo,
    pathImg: PathCI,
    title: "Continuos Integration",
    tagLine: "Harness CI is an enterprise ready, cloud native CI product.",
    valueProp:
      "An intelligent, container-native CI solution with isolated builds and standardized extensions. Build artifacts smarter and faster.",
    callout:
      "Trusted by the New York Times, Roblox, John Deere, eBay and many others to run their build and test suits balzing fast",
    intro:
      "Create a container-based build pipeline in minutes. Even testing is lightweight and fast with Test Intelligence!",
    cta: "Let's build",
    color: "#2BB1F2"
  },
  ccm: {
    logo: CCMLogo,
    pathImg: PathCCM,
    title: "Cloud Cost Management",
    tagLine: "Gain complete cloud control and automated cost optimization ",
    valueProp:
      "Achieve cost transparency and reclaim wasted spend across clouds - with built-in anomaly detection and automation",
    callout:
      "Trusted by developers and finance teams from leading startups to enterprise",
    intro:
      "Manage cloud costs and automate savings from kubernetes pods to databases",
    cta: "Connect Cloud Account",
    color: "#01C9CC"
  },
  ff: {
    logo: FFLogo,
    pathImg: PathFF,
    title: "Feature Flags",
    tagLine: "Confidently deploy features with both velocity and control.",
    valueProp:
      "Ship more features and reduce deployment risk with the control and governance enterprises require - all without breaking things, even while in production.",
    callout: "Trusted by Developers and Product teams from zero to scale.",
    intro:
      "Automate feature releases, build with flags as code, and ship more features with less risk.",
    cta: "Start Managing Features",
    color: "#EE8625"
  },
  default: {
    logo: HarnessLogo,
    pathImg: PathHarness,
    title: "Harness",
    tagLine: "Loved by Developers, Trusted by Businesses",
    valueProp:
      "Harness is the industry's first Software Delivery Platform to use AI to simplify your DevOps processes - CI, CD, Feature Flags, Cloud Costs, and much more.",
    callout:
      "Trusted by Developers, DevOps, DevSecOps, Finance and others, across world's leading digital businesses.",
    intro:
      "Define Pipelines that enable you to improve your deployment times, manage all in one place",
    cta: "Let's Deploy",
    color: "#0278D5"
  }
};

const getModuleDetails = (moduleName: string): ModuleDetail => {
  const details = MODULE_TO_DETAILS_MAP[moduleName];
  if (details) {
    return details;
  }
  return MODULE_TO_DETAILS_MAP.default;
};

export { getModuleDetails };
