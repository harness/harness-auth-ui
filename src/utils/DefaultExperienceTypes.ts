export enum DefaultExperience {
  NG = "NG",
  CG = "CG"
}
