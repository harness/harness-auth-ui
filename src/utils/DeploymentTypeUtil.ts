export const isCommunityPlan = (): boolean =>
  window.deploymentType === "COMMUNITY";

export const isOnPrem = (): boolean => window.deploymentType === "ON_PREM";

export const isSaas = (): boolean => window.deploymentType === "SAAS";
