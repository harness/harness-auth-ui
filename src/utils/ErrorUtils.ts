import toast from "react-hot-toast";

export function handleError(error: any): void {
  const response = error.data;

  if (
    response?.responseMessages?.length > 0 &&
    response?.responseMessages[response.responseMessages.length - 1].message
  ) {
    toast(
      response.responseMessages[response.responseMessages.length - 1].message ||
        ""
    );
  } else {
    toast(error.message);
  }
}
