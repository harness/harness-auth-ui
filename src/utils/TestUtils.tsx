import React from "react";
import qs from "qs";
import { compile } from "path-to-regexp";
import { Router, Route, Switch } from "react-router-dom";
import { createMemoryHistory } from "history";
import { RestfulProvider, UseGetProps } from "restful-react";
import { queryByAttribute } from "@testing-library/react";

export interface TestWrapperProps {
  path?: string;
  pathParams?: Record<string, string | number>;
  queryParams?: Record<string, unknown>;
}

export const TestWrapper: React.FC<TestWrapperProps> = (props) => {
  const { path = "/", pathParams = {}, queryParams = {} } = props;

  const search = qs.stringify(queryParams, { addQueryPrefix: true });
  const routePath = compile(path)(pathParams) + search;

  const history = React.useMemo(
    () => createMemoryHistory({ initialEntries: [routePath] }),
    []
  );

  return (
    <Router history={history}>
      <RestfulProvider base="/">
        <Switch>
          <Route exact path={path}>
            {props.children}
          </Route>
        </Switch>
      </RestfulProvider>
    </Router>
  );
};

export const queryByNameAttribute = (
  name: string,
  container: HTMLElement
): HTMLElement | null => queryByAttribute("name", container, name);

export type UseGetMockData<
  TData,
  TError = undefined,
  TQueryParams = undefined,
  TPathParams = undefined
> = Required<UseGetProps<TData, TError, TQueryParams, TPathParams>>["mock"];
