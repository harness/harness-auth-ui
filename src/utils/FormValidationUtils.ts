import { regexEmail } from "utils/StringUtils";

export function validateName(name: string): string | undefined {
  if (!name) {
    return "Name is required";
  }
}

export function validateEmail(email: string): string | undefined {
  if (!email) {
    return "An email is required";
  }

  if (!regexEmail.test(email)) {
    return "This email is not valid";
  }
}

export function validatePassword(password: string): string | undefined {
  if (!password) {
    return "A password is required";
  }

  if (password.length < 8 || password.length > 128) {
    return "The password must be between 8 and 128 characters long";
  }
}

export function validatePasswordRequiredOnly(
  password: string
): string | undefined {
  if (!password) {
    return "A password is required";
  }
}
