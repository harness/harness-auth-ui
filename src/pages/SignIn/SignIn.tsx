import React, { useEffect, useRef, useState } from "react";
import cx from "classnames";
import { Link, useHistory } from "react-router-dom";
import toast from "react-hot-toast";
import { Form } from "react-final-form";
import ReCAPTCHA from "react-google-recaptcha";

import RouteDefinitions from "RouteDefinitions";
import Field from "components/Field/Field";
import BasicLayout from "components/BasicLayout/BasicLayout";
import { useLogin } from "services/portal";

import logo from "static/images/harness-logo.svg";
import css from "./SignIn.module.css";
import AuthFooter, { AuthPage } from "components/AuthFooter/AuthFooter";
import { handleError } from "utils/ErrorUtils";
import { getAccountIdFromUrl, handleLoginSuccess } from "utils/LoginUtils";
import {
  validateEmail,
  validatePasswordRequiredOnly
} from "utils/FormValidationUtils";
import { useQueryParams } from "hooks/useQueryParams";
import { isCommunityPlan } from "utils/DeploymentTypeUtil";

const createAuthToken = (email: string, password: string): string => {
  const encodedToken = btoa(email + ":" + password);
  return `Basic ${encodedToken}`;
};

interface LoginFormData {
  email: string;
  password: string;
}

interface SignInQueryParams {
  returnUrl?: string;
  errorCode?: string;
}

const SignIn: React.FC = () => {
  const [showCaptcha, setShowCaptcha] = useState(false);
  const [captchaReponse, setCaptchaResponse] = useState<string | undefined>();
  const { mutate: login, loading } = useLogin({});
  const captchaRef = useRef<ReCAPTCHA>(null);
  const { returnUrl, errorCode } = useQueryParams<SignInQueryParams>();
  const history = useHistory();
  const accountId = returnUrl ? getAccountIdFromUrl(returnUrl) : undefined;

  // this runs once on first mount
  useEffect(() => {
    if (returnUrl) {
      // save returnUrl for SAML flow
      sessionStorage.setItem("returnUrl", returnUrl);
    } else {
      // clearing sessionStorage in case previous login was cancelled
      sessionStorage.removeItem("returnUrl");
    }

    switch (errorCode) {
      case "GATEWAY_SSO_REDIRECT_ERROR":
        toast.error(
          "Unable to sign-in using OAuth because the account is not configured with OAuth authentication."
        );
        return;
      case "unauth":
        toast.error("Current IP Address is not whitelisted.");
        return;
      case "invalidsso":
        toast.error("Invalid SSO Login.");
        return;
      case "email_verify_fail":
        toast.error(
          "Email verification failed. Please sign in to resend the email."
        );
        return;
      case "INVITE_EXPIRED":
        toast.error(
          "This invitation URL has expired. Please request for a new invitation from your admin."
        );
        return;
      case "INVITE_INVALID":
        toast.error(
          "We couldnâ€™t find an invitation matching the email address you entered. Please search your email for an invitation from Harness or contact your admin."
        );
        return;
    }
  }, []);

  const handleLogin = async (formData: LoginFormData) => {
    try {
      const response = await login(
        {
          authorization: createAuthToken(formData.email, formData.password)
        },
        {
          queryParams: {
            captcha: captchaReponse,
            accountId: accountId
          }
        }
      );
      handleLoginSuccess({
        resource: response?.resource,
        history,
        selectedAccount: accountId
      });
    } catch (error) {
      captchaRef.current?.reset();
      setCaptchaResponse(undefined);
      const errorCode = error.data?.responseMessages?.[0]?.code;
      if (errorCode === "MAX_FAILED_ATTEMPT_COUNT_EXCEEDED") {
        setShowCaptcha(true);
      }
      handleError(error);
    }
  };

  return (
    <BasicLayout>
      <div className={cx(css.signin)}>
        <div className={css.header}>
          <img src={logo} width={120} className={css.logo} alt={"Harness"} />
        </div>
        <div className={css.title}>Sign in</div>
        <div className={css.subtitle}>and get ship done.</div>
        <Form
          onSubmit={handleLogin}
          render={({ handleSubmit }) => {
            return (
              <form
                className="layout-vertical spacing-medium"
                onSubmit={handleSubmit}
              >
                <Field
                  name="email"
                  type="email"
                  label="Email"
                  placeholder="email@work.com"
                  disabled={loading}
                  validate={validateEmail}
                />
                <div
                  className="layout-vertical spacing-small"
                  style={{ position: "relative" }}
                >
                  <label htmlFor="password">Password</label>
                  <Link
                    to={RouteDefinitions.toForgotPassword()}
                    className={css.forgotLink}
                    tabIndex={-1}
                  >
                    Forgot password?
                  </Link>
                  <Field
                    name="password"
                    type="password"
                    disabled={loading}
                    validate={validatePasswordRequiredOnly}
                  />
                </div>
                {showCaptcha ? (
                  <ReCAPTCHA
                    sitekey={window.captchaToken || ""}
                    ref={captchaRef}
                    onChange={(token: string | null) => {
                      if (token) {
                        setCaptchaResponse(token);
                      }
                    }}
                  />
                ) : null}
                <input
                  type="submit"
                  value={loading ? "Signing in..." : "Sign in"}
                  className="button primary"
                  disabled={loading || (showCaptcha && !captchaReponse)}
                />
              </form>
            );
          }}
        />
        <AuthFooter
          hideOAuth={isCommunityPlan()}
          page={AuthPage.SignIn}
          accountId={accountId}
          hideSSO={isCommunityPlan()}
        />
        {window.signupExposed === "true" && (
          <div className={css.footer}>
            No account? <Link to={RouteDefinitions.toSignUp()}>Sign up</Link>
          </div>
        )}
      </div>
    </BasicLayout>
  );
};

export default SignIn;
