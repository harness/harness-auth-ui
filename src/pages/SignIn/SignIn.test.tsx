import React from "react";
import { render } from "@testing-library/react";

import SignIn from "./SignIn";
import { TestWrapper } from "utils/TestUtils";

describe("SignIn", () => {
  test("render", () => {
    const { container } = render(
      <TestWrapper path={"/signin"}>
        <SignIn />
      </TestWrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
