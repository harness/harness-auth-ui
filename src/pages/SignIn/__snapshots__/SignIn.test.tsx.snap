// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn render 1`] = `
<div>
  <div
    class="layout"
  >
    <div
      class="cardColumn"
    >
      <div
        class="card"
      >
        <div
          class="signin"
        >
          <div
            class="header"
          >
            <img
              alt="Harness"
              class="logo"
              src="test-file-stub"
              width="120"
            />
          </div>
          <div
            class="title"
          >
            Sign In
          </div>
          <div
            class="subtitle"
          >
            and get ship done.
          </div>
          <form
            class="layout-vertical spacing-medium"
          >
            <div
              class="layout-vertical spacing-small field"
            >
              <label>
                Email
              </label>
              <input
                class=""
                id="email"
                name="email"
                placeholder="email@work.com"
                type="email"
                value=""
              />
            </div>
            <div
              class="layout-vertical spacing-small"
              style="position: relative;"
            >
              <label
                for="password"
              >
                Password
              </label>
              <a
                class="forgotLink"
                href="/forgot-password"
                tabindex="-1"
              >
                Forgot Password?
              </a>
              <div
                class="layout-vertical spacing-small field"
              >
                <input
                  class=""
                  id="password"
                  name="password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <input
              class="button primary"
              type="submit"
              value="Sign In"
            />
          </form>
          <h2
            class="lineMessage"
          >
            <span
              class="message"
            >
              or login with
            </span>
          </h2>
          <div
            class="layout-horizontal spacing-auto oAuthIcons"
          >
            <a
              class="iconContainer"
              href="http://localhost/gateway/api/users/oauth2Redirect?provider=github"
            >
              <img
                height="24"
                src="test-file-stub"
                width="24"
              />
            </a>
            <a
              class="iconContainer"
              href="http://localhost/gateway/api/users/oauth2Redirect?provider=bitbucket"
            >
              <img
                height="24"
                src="test-file-stub"
                width="24"
              />
            </a>
            <a
              class="iconContainer"
              href="http://localhost/gateway/api/users/oauth2Redirect?provider=gitlab"
            >
              <img
                height="24"
                src="test-file-stub"
                width="24"
              />
            </a>
            <a
              class="iconContainer"
              href="http://localhost/gateway/api/users/oauth2Redirect?provider=linkedin"
            >
              <img
                height="24"
                src="test-file-stub"
                width="24"
              />
            </a>
            <a
              class="iconContainer"
              href="http://localhost/gateway/api/users/oauth2Redirect?provider=google"
            >
              <img
                height="24"
                src="test-file-stub"
                width="24"
              />
            </a>
            <a
              class="iconContainer"
              href="http://localhost/gateway/api/users/oauth2Redirect?provider=azure"
            >
              <img
                height="24"
                src="test-file-stub"
                width="24"
              />
            </a>
          </div>
          <button
            class="button ssoButton"
          >
            <span
              class="text"
            >
              <img
                class="icon"
                height="24"
                src="test-file-stub"
                width="24"
              />
              Single Sign-On
            </span>
          </button>
          <button
            class="button ssoButton"
          >
            <span
              class="text"
            >
              Harness Local Login
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="imageColumn"
    >
      <img
        alt=""
        aria-hidden="true"
        class="image"
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;
